spectrum/xpub.py

```diff
@@ -0,0 +1,65 @@
+from spectrum import logger
+
+
+LOGGER = logger.logger(__name__)
+
+def _log(message, *args, **kwargs):
+    LOGGER.info(message, extra={'app':'elife-xpub'}, *args, **kwargs)
+
+class PageObject:
+    def __init__(self, driver):
+        self._driver = driver
+
+    def _input(self, css_selector, text, name=None):
+        input_element = self._driver.find_element_by_css_selector(css_selector)
+        if name:
+            _log("Found %s input %s", name, css_selector)
+        input_element.send_keys(text)
+
+
+class XpubJavaScriptSession:
+    CSS_LOGIN_BUTTON = 'button[data-test-id="login"]'
+    CSS_PROFILE_MENU = 'button[data-test-id="profile-menu"]'
+
+    def __init__(self, driver):
+        self._driver = driver
+
+    def login(self):
+        login_button = self._driver.find_element_by_css_selector(self.CSS_LOGIN_BUTTON)
+        _log("Found login button %s `%s`", self.CSS_LOGIN_BUTTON, login_button.text)
+        login_button.click()
+        _log("Clicked login button %s", self.CSS_LOGIN_BUTTON)
+        profile_menu = self._driver.find_element_by_css_selector(self.CSS_PROFILE_MENU)
+        _log("Found profile menu %s", self.CSS_PROFILE_MENU)
+        profile_menu.click()
+        _log("Clicked profile menu %s", self.CSS_PROFILE_MENU)
+
+    def dashboard(self):
+        dashboard_button = self._driver.find_element_by_link_text('Dashboard')
+        dashboard_button.click()
+        return XpubDashboardPage(self._driver)
+
+
+class XpubDashboardPage(PageObject):
+    CSS_NEW_SUBMISSION_BUTTON = 'button[data-test-id="desktop-new-submission"]'
+
+    def create_initial_submission(self):
+        new_submission_button = self._driver.find_element_by_css_selector(self.CSS_NEW_SUBMISSION_BUTTON)
+        new_submission_button.click()
+        return XpubInitialSubmissionAuthorPage(self._driver)
+
+
+class XpubInitialSubmissionAuthorPage(PageObject):
+    CSS_INPUT_FIRST_NAME = 'input[name="author.firstName"]'
+    CSS_INPUT_LAST_NAME = 'input[name="author.lastName"]'
+    CSS_INPUT_EMAIL = 'input[name="author.email"]'
+    CSS_INPUT_AFFILIATION = 'input[name="author.aff"]'
+    CSS_NEXT = 'button[data-test-id="next"]'
+
+    def next(self):
+        self._input(self.CSS_INPUT_FIRST_NAME, 'Josiah', 'first name')
+        self._input(self.CSS_INPUT_LAST_NAME, 'Carberry', 'last name')
+        self._input(self.CSS_INPUT_EMAIL, 'j.carberry@example.com', 'email')
+        self._input(self.CSS_INPUT_AFFILIATION, 'Brown University', 'affiliation')
+        next_button = self._driver.find_element_by_css_selector(self.CSS_NEXT)
```

Comment:
In general, I wouldn't expect a sub class to access "private" members of the base class.

Reply:
Yes, this should be a `protected` member, visible in subclasses but not publicly. Looking for a convention.

Reply:
PEP 8 says it's a weak "internal use" indicator so not necessarily class-private (can be used at module level too), there is nothing more specific for `protected` visibility which is not a thing in Python

Reply:
The double underscore with name mangling is rarely used in Python. I'm not sure whether one would argue that the single underscore is more akin to _private_ or _protected_. In the "We're all adults here" spirit, fields are probably more generously made public in general. Not much a fan of _protected_ anyway, either it's part of the API or not. Back in Java I had to sometimes subclass just to get access to fields which might indicate a design flaw. The name doesn't really matter so much.

I guess the important point is that maybe we shouldn't tie sub classes to strongly to the driver. But then again if we are not going to have that many tests this doesn't matter so much either.

Reply:
I think some of the code reuse could be achieved with object composition, but will have to see more cases before it becomes obvious how